### ISSUE TITLE
[feature-request] Have SuperTextField work with TapRegion

### ISSUE THREAD

**Issue by jellynoone** [superlistapp/super_editor#1263]
Currently, it is not possible to easily control the focus of a `SuperTextField` using a `TapRegion` since the handles and tooltip do not use a `TapRegion`. Would it be possible to allow the user of a `SuperTextField` to set a `TapRegion.groupId` for the handles and all the different widgets displayed in the overlay, thereby making possible to easily control the focus, such as:
```dart
Widget build(BuildContext context) {
  return TapRegion(
      groupId: 'group-id',
      onTapInside: (_) => _focusNode.requestFocus(),
      onTapOutside: (_) => _focusNode.unfocus(),
      child: SuperTextField(
        tapRegionGroupId: 'group-id',
        focusNode: _focusNode,
        // additional properties
      )
    );
}
```

In other words, flutter includes a `TextFieldTapRegion` for its selection handles and the selection toolbar, which enables a developer to use a `TextFieldTapRegion` to expand the focus of a text field to related fields or to close it preemptively. The example above might not necessarily be the best one. But I was wondering if, in general, exposing these settings would make sense for the project or not?

**Comment by matthew-carroll** [comment]
I'm not sure I understand the UX goals of this request. Can you describe the user behavior that isn't currently supported by SuperTextField? How does this concept of a tap region manifest itself in the user experience?

If the goal of this ticket is about developer programmatic control, and not the user experience, can you describe the specific technical goal that a developer has? Focus is mentioned a number of times, but it's not clear to me which specific focus behavior you're trying to use, which SuperTextField doesn't support.

**Comment by jellynoone** [comment]
Hi, @matthew-carroll, thank you for responding.

So currently, what I'm trying to remove focus from the SuperTextField, anytime, anything that isn't part of the field is tapped on. However, tapping on the selection toolbar or the selection handles should not remove that focus as those are conceptually part of the field.

The easiest way I thought of implementing this was using the TapRegion, similar to how it can be done with EditableText.

When trying to use this solution, however, I noticed a problem, which is that when selection handles were tapped on, the focus was removed. The code I ran was something like this:
```dart
TapRegion(
  onTapOutside: (_) => _focusNode.unfocus(),
  onTapInside: (_) => _focusNode.requestFocus(),
  child: SuperTextField(
    focusNode: _focusNode,
    // some more properties
   )
);
```

The problem, here is that the default selection handles and toolbar of the `SuperTextField` do not belong to any `TapRegion` group. Thereby, the `TapRegion` has no way of being connected to it. So the `onTapOutside` callback gets invoked when the selection handles are used.

My proposal here was if it were possible to have the SuperTextField selection handles and other overlay driven widgets be wrapped in a `TapRegion` with either a configurable `groupId`, (this would evidently require the parameter to be passed through many different layers and would probably require a lot of simple but tedious work) or perhaps a simpler approach would be to follow what Flutter does with `TextFieldTapRegion`, the SuperTextField could have its own `SuperTextFieldTapRegion` with the `groupId` set to `SuperTextField` and the clients could then use this to easily expand the tap region of the `SuperTextField` without the need to make the `groupId` configurable.

In short:
**What I'm trying to achieve?** Unfocus the text field whenever the user taps on anything that doesn't belong to the text field.
**Why I raised this issue?** When looking at `SuperTextField` as an alternative to `EditableText`, I noticed it's (in my opinion) lacking in this implementation of the TapRegion, that the `EditableText`.

While I understand that my use case, might not bring the best or most intuitive user experience. I think the ability for clients of the package to be able to use TapRegions as they would with the regular EditableText from Flutter, would be desired.

**Comment by matthew-carroll** [comment]
Thanks for the additional information. I'll need to look deeper into what the Flutter team intends for people to do with tap regions. Adding those might make sense.

It also sounds like a possible description of this problem is that apparently we aren't letting you provide your own handle widgets. I'm guessing that if you could provide your own handle widgets, and toolbar widget, etc, then you could choose to wrap them in whatever you want, including a `TapRegion`.

We'll need to look into it and see what looks like the most composable and extensible approach to this issue.

**Comment by matthew-carroll** [comment]
I found the following in the `TapRegion`s documentation:

> This widget indicates to the nearest ancestor [TapRegionSurface](https://api.flutter.dev/flutter/widgets/TapRegionSurface-class.html) that the region occupied by its child will participate in the tap detection for that surface.
> 
> If this region belongs to a group (by virtue of its [groupId](https://api.flutter.dev/flutter/widgets/TapRegion/groupId.html)), all the regions in the group will act as one.
> 
> If there is no [TapRegionSurface](https://api.flutter.dev/flutter/widgets/TapRegionSurface-class.html) ancestor, [TapRegion](https://api.flutter.dev/flutter/widgets/TapRegion-class.html) will do nothing.

At first glance this seems like a problem because mobile drag handles, toolbars, magnifiers, etc., are all displayed in the app `Overlay`. They aren't descendants of the `SuperTextField`. If `TapRegion`s depend upon subtree relationships, then I don't think they'll work for us. I haven't checked Flutter's own implementation of `TextField` to see what they do about it.

We should try `TapRegion` and see if it works or not.

**Comment by jellynoone** [comment]
Thank you for getting back to me and looking into this. Looking, at where Flutter uses `TapRegionSurface` it appears to be inside the `WidgetsApp` above the navigator, thereby, the toolbars, drag handles etc. would all register to the same `TapRegionSurface` even if displayed through a different `OverlayEntry`.
I think the relationship here is that different `TapRegion`s register themselves on the `TapRegionSurface` and in a tipical application there is only one and its above the navigator, so all `TapRegion`s registers themselves on it.

**Comment by matthew-carroll** [comment]
@jellynoone I noticed that when I run the example app and I open the Android `SuperTextField`, I'm able to tap anywhere in that demo and remove focus from the text field. Is there something about that demo that doesn't apply to your app? Can you help me understand why the example app seems to work out of the box, but you're needing to build extra stuff to get the same effect?

**Comment by jellynoone** [comment]
I'll have a look and see what I can find out. Just to be clear the example is: super_editor/example/lib/demos/supertextfield/android/demo_superandroidtextfield.dart

**Comment by jellynoone** [comment]
Looking around, I think I found the difference.
In short, the demo wraps the `Scaffold` body in a `GestureDetector` that requests focus on tap.

Long version:
1. `SuperAndroidTextFieldDemo` is composed from `MobileSuperTextFieldDemo`
https://github.com/superlistapp/super_editor/blob/3fa362e2c121b33c1bb53521415390bab51e02d3/super_editor/example/lib/demos/supertextfield/android/demo_superandroidtextfield.dart#L27-L32

2. `MobileSuperTextFieldDemo` wraps the `Scaffold.body` in a `GestureDetector`

https://github.com/superlistapp/super_editor/blob/3fa362e2c121b33c1bb53521415390bab51e02d3/super_editor/example/lib/demos/supertextfield/_mobile_textfield_demo.dart#L86-L96

3. When tapped on other parts of the body, the `GestureDetector` will remove focus.

**Comment by jellynoone** [comment]
In addition, it should be noted that the way I read the example, it only works if the user TAPS on any other place on screen, however, if you'd for instance DRAG across the page, I believe the text field would maintain focus.

This is not the case with using the `TapRegion`. There you receive the `PointerDownEvent` directly and there is no differentiation between different gestures. Meaning, using the `TapRegion` one could immediately remove focus from the text field as soon as the screen is touched not after being tapped.

**Comment by matthew-carroll** [comment]
Thanks. I've prototyped with `TapRegion`s and it appears to work as desired. I'm now considering the possible API approaches in terms of how much control a developer should have over this. As it stands right now, I've wrapped the text field area in a `TapRegion`, as well as each drag handle, and the toolbar. The options are:

- Don't add any handlers inside the text field. Let users wrap the whole text field with their own `TapRegion` if they want to hook up auto-unfocus.
- Take the callback methods from `TapRegion` and add the same callbacks to `SuperTextField`, then connect them.
- Add a flag for whether the text field should auto-unfocus when tapped outside.
- Always auto-unfocus when tapped outside.

I'm leaning towards the first option because I never like to be too prescriptive. I also don't like depending too much on non-core Flutter widgets because the Flutter team has a way of screwing them up. But maybe `TapRegion` is low-level enough that they'll leave it alone.

**Comment by jellynoone** [comment]
Thank you for the response. Personally, I think the first approach is the best, however, it should be noted that the drag handlers and toolbar will have to be wrapped by the plugin, since the user doesn't have control over them.

In addition, the reason the user should be responsible for wrapping the text field is also because this way they can increase the tap area. What I mean by this is, if a user adds a label or some decoration or some padding or for instance a display of an error below the text field, these can all be wrapped in the same `TapRegion` and a click on any of these parts of the text field could than be made to trigger the focus.

In addition, I would recommend that the plugin follows a similar approach that flutter does with the `TapRegion`s in relation to the `TextField`. There they automatically provide a `groupId` special purposed for text fields via `TextFieldTapRegion`, (which is a subclass of `TapRegion` providing a `groupId` of value `EditableText`).

I would suggest the plugin, for instance, defines a `SuperTextFieldTapRegion` and uses it in its handlers and toolbar. Then the developer would simply have to use it to wrap their `SuperTextField`. This way basic differentiation between different `TapRegion`s is established and the plugin doesn't have to deal with providing users with the option to for instance define their own `groupId`.

Example:
```dart
class SuperTextFieldTapRegion extends TapRegion {
  const SuperTextFieldTapRegion({
    super.key,
    required super.child,
    super.enabled,
    super.onTapOutside,
    super.onTapInside,
    super.debugLabel,
  }) : super(groupId: SuperTextField);
}


// ...


Widget build(BuildContext context) {
  return SuperTextFieldTapRegion(
    child: Column(children: [
      Text('My text field'),
      SuperTextField(),
      Text('Some error'),
    ]),
  );
}
```

**Comment by jellynoone** [comment]
Alternatively, the plugin could just use `TextFieldTapRegion`. This way if the `SuperTextField` ever exposes builders for the toolbar and drag handles and the user uses those provided by Flutter, then those would work out of the box if they wrapped `TextFieldTapRegion` since flutter uses that. However, since the user is the one responsible for wrapping the field in the TextFieldTapRegion, they would simply then change their original `SuperTextFieldTapRegion` to `TextFieldTapRegion` if they'd use the flutter provided toolbar and drag handles.

However, it should also be noted that if the plugin uses its own `SuperTextFieldTapRegion` there could be an edge case where the user would, for instance, use the toolbar provided by Flutter and the drag handles provided by the plugin. In that case, the user would have to wrap their text field in both `TextFieldTapRegion` and `SuperTextFieldTapRegion` to ensure that both parts work.

And we could also consider the case, where the user might use their own toolbar and drag handles across their application for both the flutter provided text fields and the super text fields. In this case, the user would have to wrap each in both tap regions in order to ensure that all of their fields would work. I'm not sure if this would actually cause conflicts.

For this reason, I think just using the `TextFieldTapRegion` would be best.

**Comment by matthew-carroll** [comment]
Here's a PR: https://github.com/superlistapp/super_editor/pull/1310
